### PR TITLE
KAA-40: Add functionality to restore forgotten password.

### PR DESCRIPTION
**Pull request by ikulikov**

### CODE REVIEW

File: server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/Login.java [hunk 1]
```diff
@@ -193,6 +269,52 @@ else if (!isEmpty(error)) {
...
     }
+    
+    private void checkPasswordReset(AsyncCallback<Void> callback) {
+        if (historyParams.containsKey(UrlParams.RESET_PASSWORD)) {
+            String passwordResetHash = historyParams.get(UrlParams.RESET_PASSWORD);
+            resetCurrentHistory();
+            if (!Utils.isBlank(passwordResetHash) && passwordResetHash.length()==128) {
+                resetPassword(passwordResetHash, callback);
+            }
+            else {
```
Comment (ghost) [inline server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/Login.java]: Code style

File: server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/Login.java [hunk 2]
```diff
@@ -193,6 +269,52 @@ else if (!isEmpty(error)) {
             e.printStackTrace();
         }
     }
+    
+    private void checkPasswordReset(AsyncCallback<Void> callback) {
+        if (historyParams.containsKey(UrlParams.RESET_PASSWORD)) {
+            String passwordResetHash = historyParams.get(UrlParams.RESET_PASSWORD);
+            resetCurrentHistory();
+            if (!Utils.isBlank(passwordResetHash) && passwordResetHash.length()==128) {
```
Comment (ghost) [inline server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/Login.java]: Magic number

File: server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/Login.java [hunk 3]
```diff
@@ -193,6 +269,52 @@ else if (!isEmpty(error)) {
...
+                        public void onOk() {
+                            callback.onSuccess(null);
+                        }
+                    },
+                    Utils.constants.passwordWasReset(),
+                    Utils.messages.passwordWasReset());
+                    dialog.show();
+                    dialog.center();
+                }
+                else {
```
Comment (ghost) [inline server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/Login.java]: Code style

File: server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/mvp/view/dialog/ResetPasswordDialog.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+            }
+        });
+        addButton(sendResetPasswordLinkButton);
+        addButton(cancelButton);
+
+        sendResetPasswordLinkButton.setEnabled(false);
+    }
+
+    private boolean validate() {
+        boolean result = !isEmpty(usernameOrEmail.getValue());
```
Comment (ghost) [inline server/admin/src/main/java/org/kaaproject/kaa/server/admin/client/mvp/view/dialog/ResetPasswordDialog.java]: The 'result' variable isn't needed.

File: server/admin/src/main/java/org/kaaproject/kaa/server/admin/services/KaaAuthServiceImpl.java [hunk 5]
```diff
@@ -182,11 +171,56 @@ public ResultCode checkEmailOccupied(String email, Long userId) throws Exception
...
+        } else {
+            return ResultCode.OK;
+        }
+    }
+
+    @Transactional(propagation=Propagation.REQUIRED, rollbackFor=Exception.class)
+    public ResultCode sendPasswordResetLinkByEmail(String usernameOrEmail)
+            throws Exception {
+        User userEntity = userFacade.findByUsernameOrMail(usernameOrEmail);
+        if (userEntity == null) {
```
Comment (ghost) [inline server/admin/src/main/java/org/kaaproject/kaa/server/admin/services/KaaAuthServiceImpl.java:201]: Maybe better to use the 'checkUsernameOrEmailExists' function

File: server/admin/src/main/java/org/kaaproject/kaa/server/admin/services/KaaAuthServiceImpl.java [hunk 6]
```diff
@@ -182,11 +171,56 @@ public ResultCode checkEmailOccupied(String email, Long userId) throws Exception
...
+        userFacade.save(userEntity);
+        messagingService.sendPasswordResetLink(passwordResetHash, userEntity.getUsername(), userEntity.getMail());
+        return ResultCode.OK;
+    }
+
+    @Transactional(propagation=Propagation.REQUIRED, rollbackFor=Exception.class)
+    public ResultCode resetPasswordByResetHash(String passwordResetHash)
+            throws Exception {
+        User userEntity = userFacade.findByPasswordResetHash(passwordResetHash);
+        if (userEntity == null) {
```
Comment (ghost) [inline server/admin/src/main/java/org/kaaproject/kaa/server/admin/services/KaaAuthServiceImpl.java:214]: Maybe better to use the 'checkUsernameOrEmailExists' function

**Comment by shpigor** [comment]
Please, use code formatting for all changes.
Reviewed

**Pull request merged**